### PR TITLE
Refactor 404 and 500 approach

### PR DESCRIPTION
**Pull request by natemoo-re**
## Changes

- Refactors our base `app` to handle `404` and `500` routes more consistently
- Should now support pre-rendered `404` and `500` pages by making a new request to the origin server
- Treating as a patch, but can release as minor if we're concerned about the approach?
- Related to https://github.com/withastro/astro/issues/7735, https://github.com/withastro/astro/issues/7616

## Testing

Tests updated!

## Docs

N/A

### CODE REVIEW

**Review by matthewp — APPROVED** [review]

**Review by ematipico — APPROVED** [review]
I like this approach a lot. Great refactor!

File: packages/astro/src/core/app/index.ts [hunk 1]
```diff
@@ -307,6 +262,62 @@ export class App {
...
+
+	#mergeResponses(newResponse: Response, oldResponse?: Response) {
+		if (!oldResponse) return newResponse;
+		const { status, statusText, headers } = oldResponse;
+		return new Response(newResponse.body, {
+			status,
+			statusText,
+			headers: new Headers(Array.from(headers))
+		})
+	}
```
Comment (natemoo-re) [inline packages/astro/src/core/app/index.ts:312]: Not sure if this is totally needed? We had a test that confirmed that `status` and `statusText` from the original `Response` were preserved.

Trying to think through edge cases here—a bit unclear if headers from the `newResponse` should be preserved/have precedence?

File: packages/integrations/node/test/fixtures/prerender-404/package.json [hunk 2]
```diff
@@ -0,0 +1,9 @@
+{
+  "name": "@test/nodejs-prerender-404",
```
Comment (natemoo-re) [inline packages/integrations/node/test/fixtures/prerender-404/package.json:2]: Moved this from the `astro` package to the `@astrojs/node` package so we can test that the prerendered 404 page is properly handled in prod.

**Review by natemoo-re — COMMENTED** [review]
A few updates! Moved some tests around, added a couple small fixes

File: packages/astro/src/core/app/index.ts (hunk 1 above)
Comment (matthewp) [reply packages/astro/src/core/app/index.ts]: I'm not sure either but preserving them feels like the right thing to do...

File: packages/astro/src/core/app/index.ts [hunk 3]
```diff
@@ -307,6 +264,63 @@ export class App {
...
+		Reflect.set(response, responseSentSymbol, true);
+		return response;
+	}
+
+	#mergeResponses(newResponse: Response, oldResponse?: Response) {
+		if (!oldResponse) return newResponse;
+		const { status, statusText, headers } = oldResponse;
+
+		return new Response(newResponse.body, {
+			status: status === 200 ? newResponse.status : status,
```
Comment (ematipico) [inline packages/astro/src/core/app/index.ts:311]: This is weird; why do we return `newRespnose.status` if we check the original `staus`? Can you add a comment?

**Comment by ematipico** [comment]
@natemoo-re I left some feedback, could you please address it?

**Pull request merged**